### DISCUSSION TITLE
Extend GitLab provider to handle both self-hosted instances and additional scopes

### DISCUSSION THREAD

**Discussion by db6edr** [nextauthjs/next-auth#7323] [Ideas]
### Description 📓

The current implementation of the GitLab provider only works with the managed instance at https://gitlab.com and the scope `read_user`; see line https://github.com/nextauthjs/next-auth/blob/039a14d99261955be3b0d3433dd603f639a6efe6/packages/core/src/providers/gitlab.ts#L100 and following.

To make the provider work with a self-hosted instance, one needs to find the three options that need to be replaced and their respective values.

Similar when needing additional scopes; for those it's only one option that needs to be changed, but it must be found as well.

I suggest to extend the provider function with additional parameters
- `gitlabInstance` that defaults to `https://gitlab.com/` when omitted or falsy and
- `scopes` that defaults to `read_user`

Any recommendations, suggestions, questions? If not, I'd start a PR.

### How to reproduce ☕️

Proposed implementation:

```typescript
export default function GitLab<P extends GitLabProfile>(
  options: OAuthUserConfig<P>,
  gitlabInstance = "https://gitlab.com/",
  scopes: string|string[] = "read_user"
): OAuthConfig<P> {
  if (!gitlabInstance.endsWith("/")) {
    gitlabInstance = gitlabInstance + "/";
  }

  scopes = Array.isArray(scopes) ? scopes.join("+") : scopes.replace(" ", "+");

  authorization: `${gitlabInstance}oauth/authorize?scope=${scopes}`;
  token: `${gitlabInstance}oauth/token`;
  userinfo: `${gitlabInstance}api/v4/user`;
...
```

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

**Comment by rohan-chaturvedi** [comment]
It seems this is already supported, simply by providing your gitlab instance host in `authorization.url` as well as `token` and `userInfo` in the provider settings. We were able to easily add support for self-hosted gitlab to our app by supplying these values: https://github.com/phasehq/console/pull/259/files#diff-72917a6d5e6f09a7fd63d62c78d8117fc723f0d5409411cfde0d14a5bbda5842

**Reply by NeroBlackstone** [reply]
This functionality should be merged into the mainline of auth.js
